### ISSUE TITLE
Continuous 4byte emoji would crash when ReplaceFunc()

### ISSUE THREAD

**Issue by hachi8833** [dlclark/regexp2#5]
Hello, it's been a long time.

Today I found an issue regarding some special "4byte" emojis on ReplaceFunc().

* sample 4byte emojis: 📍😏️📣🍣🍺
* sample 3byte emoji: ✔️⚾️

You can inspect the above with http://r12a.github.io/apps/conversion/ like the following:

![image](https://cloud.githubusercontent.com/assets/4289625/22862574/ec5c9334-f174-11e6-829f-e27d4afc955c.png)

## Sample1: causes panic

Please take a look at the following: You can reproduce the issue by uncommenting the `str` assignment lines one by one.

As far as I checked, ReplaceFunc()'d get panic under the following condition:
- target contains some **continuous 4byte emojis**, and
- regex contains **3bytes UTF-8** characters and contains **NO 4byte emojis**

```go
package main

import (
	"github.com/dlclark/regexp2"
	"github.com/k0kubun/pp"
)

func main() {
	str := "高" // panic: Japanese Kanji
	// str := "は" // panic: Japanese Hiragana
	// str := "パ" // panic: Japanese Katakana
	// str := "[a-zA-Z0-9]{,2}" // works fine: Japanese Hiragana
	// str := "峰起|烽起" // works fine: longer Japanese Hiragana (I wonder why)
	// str := "フトレス" // panic: longer Japanese Katakana
	// str := "ALLWAYS|Allways|allways|AllWays" // works fine: Alphabet
	// str := "📍" // works fine: 4byte emoji
	// str := "📍📍" // works fine: continuous 4byte emoji
	// str := "✔️" // panic: 3byte emoji
	// str := "✔️✔️" // panic: coutinuous 3byte emoji
	// str := "📍️✔️" // works fine: 4 and 3byte emoji
	// str := "️✔📍️" // works fine: 3 and 4byte emoji
	// str := "📍️は️" // works fine: 4byte emoji and Hiragana
	// str := "️は📍️" // works fine: Hiragana and 4byte emoji

	re := regexp2.MustCompile(str, 0)
	result, _ := re.ReplaceFunc("📍✔️😏⚾️📣🍣🍺🍺 <- continuous 4byte emoji 寿司ビール文字あり", func(m regexp2.Match) string {
		return "࿗" + "࿘" + string(m.Capture.Runes()) + "࿌"
	}, -1, -1)

	pp.Println(result)
}
```

## Sample2: all works fine

The following is a kind of control group that works fine. The key is that the target contains no "continuous 4byte emojis".

```go
package main

import (
	"github.com/dlclark/regexp2"
	"github.com/k0kubun/pp"
)

func main() {
        // All of the following patterns work fine perhaps because ""✔✔⚾⚾️ <- 3byte emoji 寿司ビール文字なし" contains no continuous 4byte emojis. You can check them by uncommenting them one by one.
	str := "高"
	// str := "は"
	// str := "パ"
	// str := "[a-zA-Z0-9]{,2}"
	// str := "峰起|烽起"
	// str := "フトレス"
	// str := "ALLWAYS|Allways|allways|AllWays"
	// str := "📍" 
	// str := "📍📍" 
	// str := "✔️" 
	// str := "✔️✔️" 
	// str := "📍️✔️" 
	// str := "️✔📍️" 
	// str := "📍️は️" 
	// str := "️は📍️" 

	re := regexp2.MustCompile(str, 0)
       // The following target works fine: there's no continuous 4byte emojis
	result, _ := re.ReplaceFunc("✔✔⚾⚾️ <- 3byte emoji 寿司ビール文字なし", func(m regexp2.Match) string {
		return "࿗" + "࿘" + string(m.Capture.Runes()) + "࿌"
	}, -1, -1)

	pp.Println(result)
}
```

### FYI

The issue looks a little bit similar to "sushi-beer" issue: https://gist.github.com/kamipo/37576ce436c564d8cc28

I hope you'd check and fix it.

Best regards, 🙇

**Comment by dlclark** [comment]
Thank so for the great big report! I'll look into it this afternoon (US Central time) and get back to you.

**Comment by dlclark** [comment]
Ok, I figured it and out.  It will panic if all of these are true:  
- your pattern has a fixed text prefix 
- the prefix has at least one character between 0x80 and 0xFFFF
- the prefix has no characters more than 2-bytes
- you try to match against text that has unicode characters that are longer than 2-bytes


The Boyer-Morre algorithm I had implemented (from .NET) doesn't support more than 2-byte unicode chars, so in the past I added a fix so it isn't used if the _pattern_ has extended characters.  However, I forgot to handle the case if the pattern didn't have them, but the input text has them.  

I've now added that support and a test for it.

Thanks again for reporting the issue.
